tools/bug-report/pkg/bugreport/bugreport.go

```diff
@@ -0,0 +1,262 @@
+	}
+	resources, err := cluster2.GetClusterResources(context.Background(), clientset)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Cluster resource tree:\n\n%s\n\n", resources)
+	paths, err := filter.GetMatchingPaths(config, resources)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Fetching logs for the following containers:\n\n%s\n", strings.Join(paths, "\n"))
+
+	gatherInfo(config, resources, paths)
+	if len(gErrors) != 0 {
+		log.Errora(gErrors.ToError())
+	}
+
+	// TODO: sort by importance and discard any over the size limit.
+	for path, text := range logs {
+		namespace, _, pod, _, err := cluster2.ParsePath(path)
+		if err != nil {
+			log.Errorf(err.Error())
+			continue
+		}
+		writeFile(archive.ProxyLogPath(tempDir, namespace, pod), text)
+	}
+	return nil
+}
+
+// gatherInfo fetches all logs, resources, debug etc. using goroutines.
+// proxy logs and info are saved in logs/stats/importance global maps.
+// Errors are reported through gErrors.
+func gatherInfo(config *config.BugReportConfig, resources *cluster2.Resources, paths []string) {
+	// no timeout on mandatoryWg.
+	var mandatoryWg sync.WaitGroup
+	cmdTimer := time.NewTimer(time.Duration(config.CommandTimeout))
+
+	clusterDir := archive.ClusterInfoPath(tempDir)
+	getFromCluster(content.GetK8sResources, &content.Params{DryRun: config.DryRun}, clusterDir, &mandatoryWg)
+	getFromCluster(content.GetCRs, &content.Params{DryRun: config.DryRun}, clusterDir, &mandatoryWg)
+	getFromCluster(content.GetEvents, &content.Params{DryRun: config.DryRun}, clusterDir, &mandatoryWg)
+	getFromCluster(content.GetClusterInfo, &content.Params{DryRun: config.DryRun}, clusterDir, &mandatoryWg)
+	getFromCluster(content.GetSecrets, &content.Params{DryRun: config.DryRun, Verbose: config.FullSecrets}, clusterDir, &mandatoryWg)
+	getFromCluster(content.GetDescribePods, &content.Params{DryRun: config.DryRun, Namespace: config.IstioNamespace}, clusterDir, &mandatoryWg)
+
+	// optionalWg is subject to timer.
+	var optionalWg sync.WaitGroup
+	for _, p := range paths {
+		namespace, _, pod, container, err := cluster2.ParsePath(p)
+		if err != nil {
+			log.Error(err.Error())
+			continue
+		}
+
+		switch {
+		case container == "istio-proxy":
+			getFromCluster(content.GetCoredumps, &content.Params{DryRun: config.DryRun, Namespace: namespace, Pod: pod, Container: container},
+				archive.ProxyCoredumpPath(tempDir, namespace, pod), &mandatoryWg)
+			getProxyLogs(config, resources, p, namespace, pod, container, &optionalWg)
+
+		case strings.HasPrefix(pod, "istiod-") && container == "discovery":
+			getFromCluster(content.GetIstiodInfo, &content.Params{DryRun: config.DryRun, Namespace: namespace, Pod: pod, Container: container},
+				archive.IstiodPath(tempDir, namespace, pod), &mandatoryWg)
+			getIstiodLogs(config, resources, namespace, pod, &mandatoryWg)
+
+		}
+	}
+
+	// Not all items are subject to timeout. Proceed only if the non-cancellable items have completed.
+	mandatoryWg.Wait()
```

Comment:
is it possible this will hang forever?

Reply:
yes, good point. this needs an internal timeout of some kind.

Reply:
done